### PR TITLE
Skip missing events

### PR DESCRIPTION
**Pull request by garemoko**
Re-do of #30. Under construction.

### CODE REVIEW

File: src/Controller.php [hunk 1]
```diff
@@ -76,8 +61,13 @@ public function createEvents(array $events) {
...
             $route = isset($opts['eventname']) ? $opts['eventname'] : '';
             if (isset(static::$routes[$route]) && ($opts['userid'] > 0 || $opts['relateduserid'] > 0)) {
-                $event = '\LogExpander\Events\\'.static::$routes[$route];
-                array_push($results , (new $event($this->repo))->read($opts));
+                try {
+                    $event = '\LogExpander\Events\\'.static::$routes[$route];
+                    array_push($results , (new $event($this->repo))->read($opts));
+                }
+                catch {
+                    // Error processing event; skip it.
```
Comment (garemoko) [inline src/Controller.php]: @ryansmith94 @jlowe64 any thoughts on if it's possible to hook into error logging here?

**Comment by garemoko** [comment]
@ryansmith94 is this along the lines you were thinking?

**Comment by garemoko** [comment]
@ryansmith94 another question for you; how do i go about testing the controller?

File: src/Controller.php (hunk 1 above)
Comment (ryansmith94) [inline src/Controller.php]: No idea, would be a question for @jlowe64. If it's possible, then I'd suggest creating a method in the repo to log errors and using that from the controller so that it's testable.

**Comment by ryansmith94** [comment]
> is this along the lines you were thinking?

Yes except I think it might be better to throw the exceptions in the `readStoreRecord` and `readStoreRecords` methods instead because then you don't have to throw the same error in so many different places in the code.

> how do i go about testing the controller?

That's kind of what the integration tests are good for in the plugin.

**Comment by garemoko** [comment]
Thanks. This should be good to merge then as I'll need to do the tests as a separate PR to the main repo.

**Comment by garemoko** [comment]
We can also add error logging as a future, less urgent PR

File: src/Repository.php [hunk 2]
```diff
@@ -34,6 +37,9 @@ protected function readStoreRecord($type, array $query) {
      */
     protected function readStoreRecords($type, array $query) {
         $model = $this->store->get_records($type, $query);
+        if ($model === false) {
+            throw new Exception('Object not found');
```
Comment (ryansmith94) [inline src/Repository.php]: Sorry to be a pain in the ass @garemoko, but can we change this to 'Objects not found' so it's obvious we were attempting to read multiple records? I think this might even return an empty array in such a case.

Comment (ryansmith94) [inline src/Repository.php]: Might also be better to use the word "Record" instead of "Object".

Comment (garemoko) [inline src/Repository.php]: Confirmed; this does return an empty array

Comment (ryansmith94) [inline src/Repository.php]: Ok so we can remove this error then and just return the empty array.

Comment (garemoko) [inline src/Repository.php]: Looking and thinking about this in more detail, I don;t think we should error in these cases anyway. E.g. if you have data about a quiz and the quiz still exists but the questions do not then you can still send a statement about the quiz as a whole. Conversely, it's not possible to calculate the score for a Feedback without the questions. 

TL;DR - an empty array here needs to be handled on an individual basis, probably in the translator.

Comment (ryansmith94) [inline src/Repository.php]: Yeah I agree that empty arrays shouldn't be handled here and I'm happy for that to be handled in the translator.

**Comment by garemoko** [comment]
Ok. I've more carefully tested this now. It will skip over events if it's missing data without choking. For feedback events, the feedback submitted and other question events still get sent if a question is missing. For some reason the quiz sends no statements at all if one question is missing; that's probably fixable but i don't plan to do that as part of this PR.

**Comment by ryansmith94** [comment]
Happy for this to be merged now @jlowe64.
